File: src/main/java/xyz/nkomarn/composter/network/Connection.java
package xyz.nkomarn.composter.network;

import io.netty.channel.Channel;
import net.kyori.adventure.text.Component;
import org.jetbrains.annotations.Nullable;
import xyz.nkomarn.composter.entity.Player;
import xyz.nkomarn.composter.network.protocol.ConnectionState;
import xyz.nkomarn.composter.network.protocol.handler.DefaultPacketHandler;
import xyz.nkomarn.composter.network.protocol.handler.PacketHandler;
import xyz.nkomarn.composter.network.protocol.packet.Packet;
import xyz.nkomarn.composter.network.protocol.packet.clientbound.ClientboundDisconnectPacket;
import xyz.nkomarn.composter.server.MinecraftServer;

public class Connection {

    private final MinecraftServer server;
    private final Channel channel;
    private final PacketHandler packetHandler;
    private ConnectionState state;
    private Player player;

    public Connection(MinecraftServer server, Channel channel) {
        this.server = server;
        this.channel = channel;
        this.packetHandler = new DefaultPacketHandler(server, this);
        this.state = ConnectionState.HANDSHAKING;
    }

    public ConnectionState state() {
        return state;
    }

    public PacketHandler packetHandler() {
        return packetHandler;
    }

    public void setState(ConnectionState state) {
        this.state = state;
    }

    @Nullable
    public Player player() {
        return player;
    }

    public void bindPlayer(Player player) {
        this.player = player;
    }

    public void sendPacket(Packet packet) {
        channel.writeAndFlush(packet);
    }

    public void disconnect(Component reason) {
        sendPacket(new ClientboundDisconnectPacket(reason));
    }

    public void handleDisconnection() {
        if (player != null) {
            server.playerList().playerDisconnected(player);
        }
    }

    public void close() {
        channel.closeFuture();
    }
}

File: src/main/kotlin/kyta/composter/network/Connection.kt
package kyta.composter.network

import io.netty.channel.Channel
import io.netty.channel.ChannelFuture
import io.netty.channel.ChannelFutureListener
import io.netty.util.AttributeKey
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kyta.composter.protocol.ConnectionState
import kyta.composter.protocol.Packet
import kyta.composter.protocol.PacketHandler
import kyta.composter.protocol.packet.GenericDisconnectPacket
import kyta.composter.server.MinecraftServer
import kyta.composter.world.entity.Player
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer

data class Connection(
    private val server: MinecraftServer,
    private val channel: Channel,
    var state: ConnectionState,
) : CoroutineScope {
    override val coroutineContext = Dispatchers.Unconfined
    lateinit var packetHandler: PacketHandler
    lateinit var player: Player

    fun sendPacket(packet: Packet): ChannelFuture {
        return channel.writeAndFlush(packet)
    }

    @Deprecated("Replaced by modern component API.")
    fun disconnect(message: String) {
        disconnect(Component.text(message))
    }

    fun disconnect(message: Component) {
        // todo - net refactor
        val string = LegacyComponentSerializer.legacySection().serialize(message)
        channel.writeAndFlush(GenericDisconnectPacket(string)).addListener(ChannelFutureListener.CLOSE)
    }

    fun close() {
        channel.close()
    }

    fun handleDisconnection() {
        if (this::player.isInitialized) {
            server.playerList.playerDisconnected(player)
        }
    }

    companion object {
        val KEY_PLAYER_CONNECTION: AttributeKey<Connection> = AttributeKey.valueOf("player_connection")
    }
}
